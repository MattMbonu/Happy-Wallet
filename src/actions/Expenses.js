import uuid from "uuid";
import database from "../firebase/firebase";

export const addExpense = ({
  id,
  description = "",
  note = "",
  amount = 0,
  createdAt = 0
} = {}) => ({
  type: "ADD_EXPENSE",
  expense: {
    id,
    description,
    note,
    amount,
    createdAt
  }
});

export const setExpenses = () => {
  const expenses = [];
  return dispatch => {
    return database
      .ref("expenses")
      .once("value")
      .then(snapshot => {
        snapshot.forEach(snapshotChild => {
          expenses.push({
            id: snapshotChild.key,
            ...snapshotChild.val()
          });
        });
        dispatch(setExpenseStore(expenses));
      });
  };
};

export const setExpenseStore = (expenses = []) => ({
  type: "SET_EXPENSE",
  expenses
});

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    database
      .ref("expenses")
      .push(expense)
      .then(ref =>
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        )
      );
  };
};
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  removeId: id
});

export const startRemoveExpense = ({ id } = {}) => {
  return dispatch => {
    database
      .ref(`expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
  };
};

export const editExpense = (id, updates = {}) => ({
  type: "EDIT_EXPENSE",
  updates,
  id
});
