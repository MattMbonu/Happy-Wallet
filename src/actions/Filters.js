export const setTextFilters = (text = "") => ({
  type: "SET_TEXTFILTER",
  text
});

export const sortByDate = () => ({ type: "SORT_BY_DATE", sortBy: "date" });
export const sortByAmount = () => ({
  type: "SORT_BY_AMOUNT",
  sortBy: "amount"
});
export const setStartDate = date => ({ type: "SET_START_DATE", date });
export const setEndDate = date => ({ type: "SET_END_DATE", date });
