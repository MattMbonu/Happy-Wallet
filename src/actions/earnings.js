import database from "../firebase/firebase";

const addExpense = ({
  id,
  description = "",
  amount = 0,
  createdAt = 0,
  note = ""
}) => ({
  type: "ADD_EARNING",
  earning: { id, description, amount, createdAt, note }
});

export const startAddEarning = ({
  description = "",
  amount = 0,
  createdAt = 0,
  note = ""
}) => {
  return (dispatch, getState) => {
    const earning = { description, amount, createdAt, note };
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/earnings`)
      .push(earning)
      .then(ref => dispatch(addExpense({ id: ref.key, ...earning })));
  };
};

const setEarnings = earnings => ({ type: "SET_EARNING", earnings });

export const startSetEarnings = () => {
  return (dispatch, getState) => {
    const earnings = [];
    const uid = getState().auth.uid;
    database
      .ref(`users/${uid}/earnings`)
      .once("value")
      .then(snapshot => {
        snapshot.forEach(snapshotChild => {
          console.log(snapshotChild);
          earnings.push({
            id: snapshotChild.key,
            ...snapshotChild.val()
          });
        });
      })
      .then(() => dispatch(setEarnings(earnings)));
  };
};
