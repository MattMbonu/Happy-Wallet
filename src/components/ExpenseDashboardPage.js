import React from "react";
import ConnectedExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";

const ExpenseDashboardPage = () => (
  <div>
    <p>this is my expenses Dashboard</p>
    <ExpenseListFilters />
    <ConnectedExpenseList />
  </div>
);

export default ExpenseDashboardPage;
