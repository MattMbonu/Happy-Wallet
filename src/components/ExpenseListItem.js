import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

export const ExpenseListItem = props => {
  return (
    <div>
      <Link to={`/edit/${props.id}`}>
        <h1>{props.description}</h1>
      </Link>
      <h2>
        {numeral(props.amount / 100).format("$0,0.00")}- Date:
        {moment(props.createdAt).format("MMMM Do, YYYY")}
      </h2>
    </div>
  );
};

export default connect()(ExpenseListItem);
