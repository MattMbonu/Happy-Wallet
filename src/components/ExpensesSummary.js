import React from "react";
import numeral from "numeral";
import { connect } from "react-redux";
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from "../selectors/expenses-total";

const ExpensesSummary = props => (
  <div>
    <h1>
      Viewing: {props.expenses.length}{" "}
      {props.expenses.length === 1 ? "expense" : "expenses"} totalling{" "}
      {numeral(props.expensesTotal / 100).format("$0,0.00")}
    </h1>
  </div>
);

const mapStateToProps = state => {
  const selectedExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenses: selectedExpenses,
    expensesTotal: selectExpensesTotal(selectedExpenses)
  };
};
export default connect(mapStateToProps)(ExpensesSummary);
