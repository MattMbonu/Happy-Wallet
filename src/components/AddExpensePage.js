import React, { Component } from "react";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/Expenses";
import { connect } from "react-redux";

export class AddExpensePage extends Component {
  onSubmit = expense => {
    this.props.addExpense(expense);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h1>This is my Create Expense Page page!!</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispactchToProps = dispatch => {
  return {
    addExpense: expense => dispatch(addExpense(expense))
  };
};
export default connect(
  undefined,
  mapDispactchToProps
)(AddExpensePage);
