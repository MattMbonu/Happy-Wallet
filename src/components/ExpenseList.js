import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";
import { Link } from "react-router-dom";

export const ExpenseList = props => (
  <div>
    {props.expenses.length === 0 ? (
      <p>
        no expenses click{" "}
        <Link to="/create" exact={true}>
          here
        </Link>{" "}
        to create a new expense
      </p>
    ) : (
      props.expenses.map(expense => (
        <ExpenseListItem
          key={expense.id}
          description={expense.description}
          amount={expense.amount}
          createdAt={expense.createdAt}
          id={expense.id}
        />
      ))
    )}
  </div>
);

const mapStateToProps = state => ({
  expenses: selectExpenses(state.expenses, state.filters)
});
export default connect(mapStateToProps)(ExpenseList);
