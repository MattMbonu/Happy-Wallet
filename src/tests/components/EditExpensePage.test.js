import React from "react";
import { EditExpensePage } from "../../components/EditExpensePage";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";
import { editExpense } from "../../actions/Expenses";

test("should render EditExpensePage", () => {
  const history = { push: jest.fn() };
  const editExpense = jest.fn();
  const removeExpense = jest.fn();
  const wrapper = shallow(
    <EditExpensePage
      editExpense={editExpense}
      removeExpense={removeExpense}
      expense={expenses[0]}
      history={history}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test("remove button removes expense with correct id", () => {
  const history = { push: jest.fn() };
  const editExpense = jest.fn();
  const removeExpense = jest.fn();
  const wrapper = shallow(
    <EditExpensePage
      editExpense={editExpense}
      removeExpense={removeExpense}
      expense={expenses[0]}
      history={history}
    />
  );
  wrapper.find("button").simulate("click");
  expect(removeExpense).toHaveBeenLastCalledWith(expenses[0].id);
  expect(history.push).toHaveBeenLastCalledWith("/");
});

test("onSubmit works correctly", () => {
  const history = { push: jest.fn() };
  const editExpense = jest.fn();
  const removeExpense = jest.fn();
  const wrapper = shallow(
    <EditExpensePage
      editExpense={editExpense}
      removeExpense={removeExpense}
      expense={expenses[0]}
      history={history}
    />
  );
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[1]);
  expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith("/");
});
