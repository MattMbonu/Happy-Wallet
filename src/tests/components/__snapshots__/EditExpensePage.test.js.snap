// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <h1>
    This is my edit page namaste!!
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 10000,
        "createdAt": "1970-01-01T00:00:00.000Z",
        "description": "rent",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
